README.md
# headers-utils

[![deno land](http://img.shields.io/badge/available%20on-deno.land/x-lightgrey.svg?logo=deno)](https://deno.land/x/headers_utils)
[![GitHub release (latest by date)](https://img.shields.io/github/v/release/httpland/headers-utils)](https://github.com/httpland/headers-utils/releases)
[![codecov](https://codecov.io/github/httpland/headers-utils/branch/main/graph/badge.svg)](https://codecov.io/gh/httpland/headers-utils)
[![GitHub](https://img.shields.io/github/license/httpland/headers-utils)](https://github.com/httpland/headers-utils/blob/main/LICENSE)

[![test](https://github.com/httpland/headers-utils/actions/workflows/test.yaml/badge.svg)](https://github.com/httpland/headers-utils/actions/workflows/test.yaml)
[![NPM](https://nodei.co/npm/@httpland/headers-utils.png?mini=true)](https://nodei.co/npm/@httpland/headers-utils/)

Headers utility collection.

## equalsHeaders

Check two `Headers` field name and field value equality.

```ts
import { equalsHeaders } from "https://deno.land/x/headers_utils@$VERSION/equal.ts";
import { assert, assertFalse } from "https://deno.land/std/testing/asserts.ts";

assert(equalsHeaders(new Headers({ a: "b" }), new Headers({ a: "b" })));
assertFalse(equalsHeaders(new Headers({ a: "b" }), new Headers({ c: "d" })));
```

## filterHeadersEntries

Returns a new `Headers` with all entries of the given headers except the ones
that do not match the given predicate.

```ts
import { filterHeadersEntries } from "https://deno.land/x/headers_utils@$VERSION/filter_entries.ts";
import { assert, assertFalse } from "https://deno.land/std/testing/asserts.ts";

declare const isIMFDate: (input: string) => boolean;

const headers = filterHeadersEntries(
  new Headers({
    "date": "<date>",
    "content-type": "<content-type>",
  }),
  ([key, value]) => isIMFDate(value),
);

assert(headers.has("date"));
assert(headers.has("content-type"));
```

## filterHeadersKeys

Returns a new `Headers` with all entries of the given headers except the ones
that have a key(field name) that does not match the given predicate.

```ts
import { filterHeadersKeys } from "https://deno.land/x/headers_utils@$VERSION/filter_keys.ts";
import { assert, assertFalse } from "https://deno.land/std/testing/asserts.ts";

const headers = filterHeadersKeys(
  new Headers({
    "date": "<date>",
    "content-type": "<content-type>",
  }),
  (key) => key.startsWith("content"),
);

assert(headers.has("content-type"));
assertFalse(headers.has("date"));
```

## filterHeadersValues

Returns a new `Headers` with all entries of the given headers except the ones
that have a value(field value) that does not match the given predicate.

```ts
import { filterHeadersValues } from "https://deno.land/x/headers_utils@$VERSION/filter_values.ts";
import { assert, assertFalse } from "https://deno.land/std/testing/asserts.ts";

declare const isIMFDate: (input: string) => boolean;

const headers = filterHeadersValues(
  new Headers({
    "date": "<date>",
    "content-type": "<content-type>",
  }),
  isIMFDate,
);

assert(headers.has("date"));
assertFalse(headers.has("content-type"));
```

## isHeaders

Whether the input is `Headers` or not.

```ts
import { isHeaders } from "https://deno.land/x/headers_utils@$VERSION/is.ts";
import { assert, assertFalse } from "https://deno.land/std/testing/asserts.ts";

assert(isHeaders(new Headers()));
assertFalse(isHeaders({}));
```

## stringifyHeaders

Serialize `Headers` into string.

```ts
import { stringifyHeaders } from "https://deno.land/x/headers_utils@$VERSION/stringify.ts";
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

assertEquals(stringifyHeaders(new Headers()), "");
assertEquals(
  stringifyHeaders(
    new Headers({
      "content-type": "text/plain",
      "date": "Wed, 21 Oct 2015 07:28:00 GMT",
    }),
  ),
  `content-type: text/plain
date: Wed, 21 Oct 2015 07:28:00 GMT`,
);
```

## API

All APIs can be found in the [deno doc](https://deno.land/x/headers_utils?doc).

## License

Copyright © 2023-present [httpland](https://github.com/httpland).

Released under the [MIT](./LICENSE) license


CHANGELOG.md
# 1.0.0 (2023-05-08)


### Features

* add headers utility functions ([7cfad8d](https://github.com/httpland/headers-utils/commit/7cfad8de91b6f1a2ba7c5d34eb77c7dfc5218ffc))

# 1.0.0-beta.1 (2023-05-04)


### Features

* add headers utility functions ([7cfad8d](https://github.com/httpland/headers-utils/commit/7cfad8de91b6f1a2ba7c5d34eb77c7dfc5218ffc))


is_test.ts
// Copyright 2023-latest the httpland authors. All rights reserved. MIT license.

import { isHeaders } from "./is.ts";
import { assert, assertFalse, describe, it } from "./_dev_deps.ts";

describe("isHeaders", () => {
  it("should return true if the input is headers", () => {
    assert(isHeaders(new Headers()));
  });

  it("should return false if the input is not headers", () => {
    assertFalse(isHeaders({}));
    assertFalse(isHeaders([]));
  });
});


equal_test.ts
// Copyright 2023-latest the httpland authors. All rights reserved. MIT license.

import { equalsHeaders } from "./equal.ts";
import { assert, assertFalse, describe, it } from "./_dev_deps.ts";

describe("equalsHeaders", () => {
  it("should return false", () => {
    const table: [Headers, Headers][] = [
      [new Headers({ a: "" }), new Headers()],
      [new Headers(), new Headers({ a: "" })],
      [new Headers({ a: "" }), new Headers({ a: "", b: "" })],
      [new Headers({ a: "", b: "c" }), new Headers({ a: "", d: "c" })],
      [new Headers({ a: "a" }), new Headers({ a: "b" })],
      [
        new Headers({ a: "b", b: "c" }),
        new Headers({ a: "b", b: "c", c: "d" }),
      ],
    ];

    table.forEach(([a, b]) => {
      assertFalse(equalsHeaders(a, b));
    });
  });

  it("should return true", () => {
    const table: [Headers, Headers][] = [
      [new Headers(), new Headers()],
      [new Headers({ a: "" }), new Headers({ a: "" })],
      [new Headers([["a", ""]]), new Headers({ a: "" })],
      [
        new Headers({ a: "", b: "c", c: "abc" }),
        new Headers({ a: "", b: "c", c: "abc" }),
      ],
      [
        new Headers({ a: " " }),
        new Headers({ a: "" }),
      ],
      [
        new Headers({ a: "   a " }),
        new Headers({ a: "        a       " }),
      ],
    ];

    table.forEach(([a, b]) => {
      assert(equalsHeaders(a, b));
    });
  });
});


stringify_test.ts
// Copyright 2023-latest the httpland authors. All rights reserved. MIT license.

import { stringifyHeaders } from "./stringify.ts";
import { assertEquals, describe, it } from "./_dev_deps.ts";

describe("stringifyHeaders", () => {
  it("should return string", () => {
    const table: [Headers, string][] = [
      [new Headers(), ""],
      [new Headers({ "x-test": "test" }), "x-test: test"],
      [new Headers({ "X-Test": "test" }), "x-test: test"],
      [
        new Headers({ "X-Test": "test", origin: "http://example.test" }),
        "origin: http://example.test\nx-test: test",
      ],
      [
        new Headers({ origin: "http://example.test", "X-Test": "test" }),
        "origin: http://example.test\nx-test: test",
      ],
      [
        new Headers({
          origin: "http://example.test",
          "X-Test": "test",
          "x-test": "test2",
        }),
        "origin: http://example.test\nx-test: test, test2",
      ],
      [
        new Headers({
          origin: "http://example.test",
          "X-Test": "test",
          "x-test": "test2",
        }),
        "origin: http://example.test\nx-test: test, test2",
      ],
      [
        new Headers({
          origin: "http://example.test",
          "X-Test": "test",
          "x-abc": "abc",
        }),
        "origin: http://example.test\nx-abc: abc\nx-test: test",
      ],
    ];

    table.forEach(([headers, expected]) => {
      assertEquals(stringifyHeaders(headers), expected);
    });
  });
});


equal.ts
// Copyright 2023-latest the httpland authors. All rights reserved. MIT license.
// This module is browser compatible.

/** Check two `Headers` field name and field value equality.
 *
 * ```ts
 * import { equalsHeaders } from "https://deno.land/x/headers_utils@$VERSION/equal.ts";
 * import { assert, assertFalse } from "https://deno.land/std/testing/asserts.ts";
 *
 * assert(equalsHeaders(new Headers({ a: "b" }), new Headers({ a: "b" })));
 * assertFalse(equalsHeaders(new Headers({ a: "b" }), new Headers({ c: "d" })));
 * ```
 */
export function equalsHeaders(left: Headers, right: Headers): boolean {
  const lefts = [...left];
  const rights = [...right];

  if (lefts.length !== rights.length) return false;

  for (const [key, value] of lefts.concat(rights)) {
    if (!left.has(key) || !right.has(key)) {
      return false;
    }

    if (left.get(key) !== value || right.get(key) !== value) {
      return false;
    }
  }

  return true;
}


_tools/meta.ts
import {
  BuildOptions,
  EntryPoint,
} from "https://deno.land/x/dnt@0.34.0/mod.ts";
import { expandGlobSync } from "https://deno.land/std@0.185.0/fs/expand_glob.ts";
import {
  fromFileUrl,
  join,
  parse,
  relative,
} from "https://deno.land/std@0.185.0/path/mod.ts";

interface ModuleInfo {
  readonly name: string;
  readonly path: string;
}

function path2EntryPoint(module: ModuleInfo): EntryPoint {
  const entryPoint: EntryPoint = {
    name: toRelative(module.name),
    path: toRelative(module.path),
  };

  return entryPoint;
}

function module2TypeVersions(modules: readonly ModuleInfo[]) {
  const entries = modules.map(({ name, path }) => {
    return [name, [join("types", toDts(path))]];
  });
  const map = Object.fromEntries(entries);

  return { "*": map };
}

function toRelative(path: string): string {
  return path.startsWith("./") ? path : "./" + path;
}

function toDts(path: string): string {
  return path.replace(/.ts$/, ".d.ts");
}

const root = fromFileUrl(import.meta.resolve("../"));
const entries = expandGlobSync("!(_*|*_test.ts)*.ts", {
  includeDirs: false,
  root,
});

const modules = [...entries].map(({ path }) => relative(root, path)).map(
  (path) => {
    const parsed = parse(path);
    const name = join(parsed.dir, parsed.name) + ".js";

    return { name, path };
  },
);

const entryPoints = modules.map(path2EntryPoint);
const typesVersions = module2TypeVersions(modules);

export const makeOptions = (version: string): BuildOptions => ({
  test: false,
  shims: {},
  compilerOptions: {
    lib: ["esnext", "dom", "dom.iterable"],
  },
  typeCheck: true,
  entryPoints,
  outDir: "./npm",
  package: {
    name: "@httpland/headers-utils",
    version,
    description: "Headers utility collection",
    keywords: [
      "http",
      "headers",
      "header",
      "fetch",
      "utility",
      "utilities",
      "collection",
    ],
    license: "MIT",
    homepage: "https://github.com/httpland/headers-utils",
    repository: {
      type: "git",
      url: "git+https://github.com/httpland/headers-utils.git",
    },
    bugs: {
      url: "https://github.com/httpland/headers-utils/issues",
    },
    main: undefined,
    module: undefined,
    types: undefined,
    sideEffects: false,
    type: "module",
    publishConfig: {
      access: "public",
    },
    typesVersions,
  },
  packageManager: "pnpm",
});


filter_keys.ts
// Copyright 2023-latest the httpland authors. All rights reserved. MIT license.
// This module is browser compatible.

/** Returns a new {@link Headers} with all entries of the given headers except the ones that have a key(field name) that does not match the given predicate.
 *
 * @example
 * ```ts
 * import { filterHeadersKeys } from "https://deno.land/x/headers_utils@$VERSION/filter_keys.ts";
 * import { assert, assertFalse } from "https://deno.land/std/testing/asserts.ts";
 *
 * const headers = filterHeadersKeys(
 *   new Headers({
 *     "date": "<date>",
 *     "content-type": "<content-type>",
 *   }),
 *   (key) => key.startsWith("content"),
 * );
 *
 * assert(headers.has("content-type"));
 * assertFalse(headers.has("date"));
 * ```
 */
export function filterHeadersKeys(
  headers: Headers,
  predicate: (key: string) => boolean,
): Headers {
  return new Headers([...headers].filter(([key]) => predicate(key)));
}


filter_values.ts
// Copyright 2023-latest the httpland authors. All rights reserved. MIT license.
// This module is browser compatible.

/** Returns a new `Headers` with all entries of the given headers except the ones
 * that have a value(field value) that does not match the given predicate.
 *
 * @example
 * ```ts
 * import { filterHeadersValues } from "https://deno.land/x/headers_utils@$VERSION/filter_values.ts";
 * import { assert, assertFalse } from "https://deno.land/std/testing/asserts.ts";
 *
 * declare const isIMFDate: (input: string) => boolean;
 *
 * const headers = filterHeadersValues(
 *   new Headers({
 *     "date": "<date>",
 *     "content-type": "<content-type>",
 *   }),
 *   isIMFDate,
 * );
 *
 * assert(headers.has("date"));
 * assertFalse(headers.has("content-type"));
 * ```
 */
export function filterHeadersValues(
  headers: Headers,
  predicate: (value: string) => boolean,
): Headers {
  return new Headers([...headers].filter(([_, value]) => predicate(value)));
}


filter_entries.ts
// Copyright 2023-latest the httpland authors. All rights reserved. MIT license.
// This module is browser compatible.

/** Returns a new `Headers` with all entries of the given headers except the ones
 * that do not match the given predicate.
 *
 * @example
 * ```ts
 * import { filterHeadersEntries } from "https://deno.land/x/headers_utils@$VERSION/filter_entries.ts";
 * import { assert, assertFalse } from "https://deno.land/std/testing/asserts.ts";
 *
 * declare const isIMFDate: (input: string) => boolean;
 *
 * const headers = filterHeadersEntries(
 *   new Headers({
 *     "date": "<date>",
 *     "content-type": "<content-type>",
 *   }),
 *   ([key, value]) => isIMFDate(value),
 * );
 *
 * assert(headers.has("date"));
 * assert(headers.has("content-type"));
 * ```
 */
export function filterHeadersEntries(
  headers: Headers,
  predicate: (entry: [string, string]) => boolean,
): Headers {
  return new Headers([...headers].filter(predicate));
}


stringify.ts
// Copyright 2023-latest the httpland authors. All rights reserved. MIT license.
// This module is browser compatible.

/** Special character. */
const enum Char {
  LF = "\n",
  Colon = ":",
  Space = " ",
}

/** Serialize {@link Headers} into string.
 *
 * @example
 * ```ts
 * import { stringifyHeaders } from "https://deno.land/x/headers_utils@$VERSION/stringify.ts";
 * import { assertEquals } from "https://deno.land/std/testing/asserts.ts";
 *
 * assertEquals(stringifyHeaders(new Headers()), "");
 * assertEquals(
 *   stringifyHeaders(
 *     new Headers({
 *       "content-type": "text/plain",
 *       "date": "Wed, 21 Oct 2015 07:28:00 GMT",
 *     }),
 *   ),
 *   `content-type: text/plain
 * date: Wed, 21 Oct 2015 07:28:00 GMT`,
 * );
 * ```
 */
export function stringifyHeaders(headers: Headers): string {
  return [...headers.entries()]
    .map(stringifyFieldLine)
    .join(Char.LF);
}

function stringifyFieldLine(
  entry: readonly [fieldName: unknown, fieldValue: unknown],
): string {
  return entry[0] + Char.Colon + Char.Space + entry[1];
}
